protocol/contracts/templegold/TempleGoldStaking.sol

```diff
@@ -0,0 +1,521 @@
+    ) external override whenNotPaused updateReward(staker) {
+        _getReward(staker, staker);
+    }
+
+    /**  
+     * @notice Mint and distribute Temple Gold rewards 
+     */
+    function distributeGold() external whenNotPaused {
+        _distributeGold();
+    }
+
+    /**  
+     * @notice Notify rewards distribution. Called by TempleGold contract after successful mint
+     * @param amount Amount minted to this contract
+     */
+    function notifyDistribution(uint256 amount) external override {
+        if (msg.sender != address(rewardToken) && !isElevatedAccess(msg.sender, msg.sig)) { revert CommonEventsAndErrors.InvalidAccess(); }
+        /// @notice Temple Gold contract mints TGLD amount to contract before calling `notifyDistribution`
+        nextRewardAmount += amount;
+        emit GoldDistributionNotified(amount, block.timestamp);
+    }
+    
+    /**  
+     * @notice Get reward data
+     * @return Reward data
+     */
+    function getRewardData() external override view returns (Reward memory) {
+        return rewardData;
+    }
+
+    function _getReward(address _staker, address _to) internal {
+        uint256 amount = claimableRewards[_staker];
+        if (amount > 0) {
+            claimableRewards[_staker] = 0;
+            rewardToken.safeTransfer(_to, amount);
+            emit RewardPaid(_staker, _to, amount);
+        }
+    }
+
+    function _withdrawFor(
+        address _staker,
+        address _to,
+        uint256 _amount,
+        bool _claimRewards,
+        address _rewardsTo
+    ) private updateReward(_staker) {
+        if (_amount == 0) { revert CommonEventsAndErrors.ExpectedNonZero(); }
+        uint256 _balance = _balances[_staker];
+        if (_amount > _balance) 
+            { revert CommonEventsAndErrors.InsufficientBalance(address(stakingToken), _amount, _balance); }
+        unchecked {
+            _balances[_staker] = _balance - _amount;
+        }
+        totalSupply -= _amount;
+        _moveDelegates(delegates[_staker], address(0), _amount);
+
+        stakingToken.safeTransfer(_to, _amount);
+        emit Withdrawn(_staker, _to, _amount);
+
+        // claim reward
+        if (_claimRewards) {
+            // can call internal because user reward already updated
+            _getReward(_staker, _rewardsTo);
+        }
+    }
+
+    function _earned(
+        address _account
+    ) internal view returns (uint256) {
+        return _balances[_account] * (_rewardPerToken() - userRewardPerTokenPaid[_account]) / 1e18 
+            + claimableRewards[_account];
+    }
+
+    function _applyStake(address _for, uint256 _amount) internal updateReward(_for) {
+        totalSupply += _amount;
+        _balances[_for] += _amount;
+        stakeTimes[_for] = block.timestamp;
+        emit Staked(_for, _amount);
+    }
+
+    function _rewardPerToken() internal view returns (uint256) {
+        if (totalSupply == 0) {
+            return rewardData.rewardPerTokenStored;
+        }
+
```

Comment:
some gas can be saved here when `rewardData.lastUpdateTime == rewardData.periodFinish`:

```solidity
if (rewardData.lastUpdateTime == rewardData.periodFinish) {
    return rewardData.rewardPerTokenStored;
}
```
That condition only happens when `_updateReward()` is called after the period is over. In such case, you can simply return `rewardData.rewardPerTokenStored` as no more rewards are being released, and you save reading multiple storage variables. I think this gas saving can be significant over time, because `_rewardPerToken()` will be called by multiple user facing functions, so the savings accumulate over time.

Reply:
Put in other way, if 

`rewardData.lastUpdateTime == rewardData.periodFinish`

then 

`_lastTimeRewardApplicable(rewardData.periodFinish) - rewardData.lastUpdateTime) = 0`

And therefore:

```solidity
rewardData.rewardPerTokenStored +
            (((_lastTimeRewardApplicable(rewardData.periodFinish) -
                rewardData.lastUpdateTime) *
                rewardData.rewardRate * 1e18)
                / totalSupply);`
```
becomes 

```solidity
rewardData.rewardPerTokenStored + 0;
```

so you might as well return `rewardData.rewardPerTokenStored` directly and save some gas.

Reply:
Good. I added some tests to support it.